strawberryfields/io.py

```diff
@@ -244,6 +245,66 @@ def is_free_param(param):
     return prog
 
 
+def serialize_program(prog, eng=None):
+    """Converts a Strawberry Fields program into valid Strawberry Fields code.
+
+    Args:
+        prog (Program): the Strawberry Fields program
+        eng (Engine): the Strawberryfields engine
+
+    Returns:
+        str: the Strawberry Fields code, for constructing the program, as a string
+    """
+    code_seq = ["import strawberryfields as sf"]
+
+    operations = []
+    op_imports = []
+    for cmd in prog.circuit:
+        op_name = cmd.op.__class__.__name__
+        if op_name not in op_imports:
+            op_imports.append(op_name)
+
+        if prog.type == "tdm":
+            format_dict = {k: f"p[{k[1:]}]" for k in prog.parameters.keys()}
+            op = cmd.__str__().format(**format_dict)
+        else:
+            op = cmd.__str__()
+
+        operations.append(op)
+
+    code_seq.append("from strawberryfields.ops import " + ", ".join(op_imports) + "\n")
+
+    if prog.type == "tdm":
+        code_seq.append(f"prog = sf.TDMProgram(N={prog.N})")
+    else:
+        code_seq.append(f"prog = sf.Program({prog.num_subsystems})")
+
+    if eng:
+        eng_type = eng.__class__.__name__
+        if eng_type == "RemoteEngine":
+            code_seq.append(f"eng = sf.RemoteEngine({eng.target})")
+        else:
+            if eng.backend_options:
+                code_seq.append(
+                    f"eng = sf.Engine({eng.backend_name}, backend_options={eng.backend_options})\n"
+                )
+            else:
+                code_seq.append(f"eng = sf.Engine({eng.backend_name})\n")
+
+    if prog.type == "tdm":
+        tdm_params = [f"{par}" for par in prog.tdm_params]
+        code_seq.append("with prog.context(" + ", ".join(tdm_params) + ") as (p, q):")
+    else:
+        code_seq.append("with prog.context as q:")
+
+    code_seq.append("    " + "\n    ".join(operations) + "\n")
+
+    if eng:
+        code_seq.append("results = eng.run()")
```

Comment:
`eng.run` is missing the program input, as well as other runtime options (shots, etc.)

Reply:
:scream:

Reply:
I missed adding the `prog`, but `shots` and `modes` are a bit difficult since they're actually never stored in neither the engine nor the program. They're only passed when running the engine; not while creating it. 

Either we can just leave it at `eng.run(prog)`, 

or add e.g. `shots=1` always, `eng.run(prog, shots=1)`, so that this option is clear and easy to change,

or have it as another argument in the function `generate_code(prog, eng=None, shots=1)`,

or maybe even as `generate_code(prog, eng=None, **kwargs)`, where kwargs is the same as the `engine.run` kwargs.

@josh146 Do you have any other ideas? I think I'm leaning towards the second option. :thinking:

Reply:
Oh right 🤔 

Do shots not get serialized when converting from a blackbird script to a SF program? I thought they did.

But I guess if you are starting directly from a SF program, no info will be there...

Reply:
Just had a look, and `shots` can either be part of `prog.run_options` or passed directly via `eng.run`. If passed via `eng.run` `prog.run_options` is _not_ updated. That would be easy to do, but I'm not sure if it's a good thing that running the engine would change the `shots` defined earlier in `run_options`. :thinking: